### PR TITLE
feat(evm): support Gearbox protocol in price queries

### LINKED ISSUE #7544: Support Gearbox protocol
## Abstract

We should support the Gearbox protocol.

We can start with passive earn pools and detect them.

https://app.gearbox.fi/pools

## Specification

- Properly decode all gearbox related transactions in the history events
- Detect gearbox pools and automatically find balances for them.

### PR DESCRIPTION
**Pull request by moxw**
Closes #7544

### CODE REVIEW

File: rotkehlchen/inquirer.py [hunk 1]
```diff
@@ -920,6 +924,112 @@ def find_curve_pool_price(
...
+            log.error(f'Error fetching {node_inquirer.chain_name} token while detecting underlying tokens of {token.evm_address!s}')  # noqa: E501
+            return None
+
+        underlying_token_price = self.find_usd_price(underlying_token)
+        try:
+            price_per_share = lp_contract.call(
+                node_inquirer=node_inquirer,
+                method_name='convertToAssets',
+                arguments=[1 * 10 ** (decimals := token.get_decimals())],
+            )
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:1025]:
```suggestion
                arguments=[10 ** (decimals := token.get_decimals())],
            )
```

File: rotkehlchen/inquirer.py [hunk 2]
```diff
@@ -920,6 +924,112 @@ def find_curve_pool_price(
...
+        except (RemoteError, BlockchainQueryError) as e:
+            log.error(f'Failed to query convertToAssets method in {node_inquirer.chain_name} Gearbox Pool. {e!s}')  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:1027]: add the address of the asset that was being queried

File: rotkehlchen/types.py [hunk 3]
```diff
@@ -85,7 +85,7 @@
 AERODROME_POOL_PROTOCOL = 'aerodrome_pool'
 PICKLE_JAR_PROTOCOL = 'pickle_jar'
 SPAM_PROTOCOL = 'spam'
-GEARBOX_POOL_PROTOCOL = 'gearbox_pool'
+GEARBOX_POOL_PROTOCOL = 'gearbox'
```
Comment (yabirgb) [inline rotkehlchen/types.py:88]: why this change? Doesn't it affect the assets in the asset repo?

File: rotkehlchen/inquirer.py [hunk 4]
```diff
@@ -920,6 +924,112 @@ def find_curve_pool_price(
...
+
+        if token in farming_tokens:
+            farming_contract = EvmContract(
+                address=token.evm_address,
+                abi=node_inquirer.contracts.abi('GEARBOX_FARMING_POOL'),
+                deployed_block=0,  # not used here
+            )
+            try:
+                lp_token = farming_contract.call(node_inquirer=node_inquirer, method_name='stakingToken')  # noqa: E501
+                lp_token = deserialize_evm_address(lp_token)
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:955]: the deserialize can raise `DeserializationError`. Handle it

File: rotkehlchen/inquirer.py [hunk 5]
```diff
@@ -920,6 +924,112 @@ def find_curve_pool_price(
...
+            except (RemoteError, BlockchainQueryError) as e:
+                log.error(f'Failed to query stakingToken method in Gearbox Pool. {e!s}')
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:957]: inclide the asset in the error messages otherwise we won't know what happeened

File: rotkehlchen/inquirer.py [hunk 6]
```diff
@@ -920,6 +924,112 @@ def find_curve_pool_price(
...
+            address=lp_token,
+            abi=node_inquirer.contracts.abi('GEARBOX_LP'),
+            deployed_block=0,  # not used here
+        )
+        if maybe_underlying_tokens is None or len(maybe_underlying_tokens) != 1:
+            # underlying token not recorded in the DB. Ask the chain
+            try:
+                remote_underlying_token = lp_contract.call(node_inquirer, 'underlyingToken')
+            except (RemoteError, BlockchainQueryError) as e:
+                log.error(f'Failed to query underlying token method in {node_inquirer.chain_name} Gearbox pool. {e!s}')  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:972]: this logic seems to be for the case where the asset doesn't have the underlying token right? Can you move it to its own function in the gearbox evm module? something like `ensure_gearbox_lp_underlying_tokens`. Maybe a better name. This will help clearing the logic for the reader

File: rotkehlchen/inquirer.py [hunk 7]
```diff
@@ -920,6 +924,112 @@ def find_curve_pool_price(
...
+                underlying_token = get_or_create_evm_token(
+                    userdb=node_inquirer.database,
+                    evm_address=underlying_token_address,
+                    chain_id=token.chain_id,
+                    encounter=TokenEncounterInfo(
+                        description='Detecting Gearbox pool underlying tokens',
+                    ),
+                )
+            except NotERC20Conformant as e:
+                log.error(f'Error fetching {node_inquirer.chain_name} token {underlying_token_address} while detecting underlying tokens of {token.evm_address!s}: {e!s}')  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:991]: you are not returning here so the logic will continue and error later because you don't have the underlying token

File: rotkehlchen/inquirer.py [hunk 8]
```diff
@@ -920,6 +924,112 @@ def find_curve_pool_price(
...
+                evm_address_to_identifier(
+                    address=maybe_underlying_tokens[0].address,
+                    chain_id=token.chain_id,
+                    token_type=EvmTokenKind.ERC20,
+                ),
+            )
+
+        if underlying_token is None:
+            log.error(f'Error fetching {node_inquirer.chain_name} token while detecting underlying tokens of {token.evm_address!s}')  # noqa: E501
+            return None
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:1017]: how can it be none? The logic should have failed before if it wasn't a valid gearbox token before right?

File: rotkehlchen/tests/unit/test_inquirer.py [hunk 9]
```diff
@@ -622,6 +624,27 @@ def test_find_yearn_vaults_v2_price(inquirer_defi, globaldb):
             assert result and result[0].address == underlying_token.resolve_to_evm_token().evm_address  # noqa: E501
 
 
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('should_mock_current_price_queries', [False])
+def test_find_gearbox_lp_price(inquirer: 'Inquirer', arbitrum_one_manager: 'ArbitrumOneManager'):
+    dwethv3 = EvmToken('eip155:42161/erc20:0x04419d3509f13054f60d253E0c79491d9E683399')
+    sdwethv3 = EvmToken('eip155:42161/erc20:0x6773fF780Dd38175247795545Ee37adD6ab6139a')
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_inquirer.py:631]: shouldn't this test be vcred since it does external calls? Also since you will have it vcred assert for the right price

File: rotkehlchen/tests/unit/test_inquirer.py [hunk 10]
```diff
@@ -622,6 +624,27 @@ def test_find_yearn_vaults_v2_price(inquirer_defi, globaldb):
...
+            'DELETE from underlying_tokens_list WHERE parent_token_entry=?',
+            (sdwethv3.identifier,),
+        )
+    inquirer.inject_evm_managers([(ChainID.ARBITRUM_ONE, arbitrum_one_manager)])
+    for token, underlying_token in ((sdwethv3, A_WETH_ARB), (dwethv3, A_WETH_ARB)):
+        price = inquirer.find_gearbox_price(token)
+        assert price is not None
+
+        with GlobalDBHandler().conn.read_ctx() as cursor:
+            result = GlobalDBHandler().fetch_underlying_tokens(cursor, token.identifier)
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_inquirer.py:644]:
```suggestion
            result = GlobalDBHandler.fetch_underlying_tokens(cursor, token.identifier)
```

File: rotkehlchen/types.py (hunk 3 above)
Comment (moxw) [reply rotkehlchen/types.py:88]: So this wasn't used anywhere in the code before, but the assets in the repo have been saved as protocol being `gearbox`. So, the change.

File: rotkehlchen/inquirer.py (hunk 8 above)
Comment (moxw) [reply rotkehlchen/inquirer.py:1017]: Was getting linting error because I missed the `return None` earlier, this is not needed anymore, removed.

File: rotkehlchen/chain/evm/decoding/gearbox/gearbox_cache.py [hunk 11]
```diff
@@ -199,6 +199,56 @@ def ensure_gearbox_tokens_existence(
...
+    try:
+        remote_underlying_token = lp_contract.call(node_inquirer, 'underlyingToken')
+    except (RemoteError, BlockchainQueryError) as e:
+        log.error(f'Failed to query underlyingToken method in {node_inquirer.chain_name} Gearbox pool {lp_contract.address}. {e!s}')  # noqa: E501
+        return None
+
+    try:
+        underlying_token_address = deserialize_evm_address(remote_underlying_token)
+    except DeserializationError:
+        log.error(f'underlying token call of {lp_contract.address} returned invalid address {remote_underlying_token}')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/gearbox/gearbox_cache.py:219]: add chain name in this error

File: rotkehlchen/inquirer.py [hunk 12]
```diff
@@ -920,6 +927,68 @@ def find_curve_pool_price(
 
         return (total_assets_value / total_supply) * (10 ** lp_token.get_decimals())
 
+    def find_gearbox_price(self, token: EvmToken) -> Price | None:
+        evm_manager = self.get_evm_manager(chain_id=token.chain_id)
+        node_inquirer = evm_manager.node_inquirer
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:932]: do not have variables used in only one place
```suggestion
        node_inquirer = self.get_evm_manager(chain_id=token.chain_id).node_inquirer
```

File: rotkehlchen/inquirer.py [hunk 13]
```diff
@@ -920,6 +927,68 @@ def find_curve_pool_price(
...
+            maybe_underlying_tokens = GlobalDBHandler.fetch_underlying_tokens(
+                cursor=cursor,
+                parent_token_identifier=token.identifier,
+            )
+        lp_contract = EvmContract(
+            address=lp_token,
+            abi=node_inquirer.contracts.abi('GEARBOX_LP'),
+            deployed_block=0,  # not used here
+        )
+        if maybe_underlying_tokens is None or len(maybe_underlying_tokens) != 1:
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:960]: add some comments here on what this nested if is checking. It's not clear to me.

File: rotkehlchen/inquirer.py [hunk 14]
```diff
@@ -920,6 +927,68 @@ def find_curve_pool_price(
...
+            if (
+                underlying_token := ensure_gearbox_lp_underlying_tokens(
+                    token_identifier=token.identifier,
+                    node_inquirer=node_inquirer,
+                    lp_contract=lp_contract,
+                )
+            ) is None:
+                return None
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:968]: Isn't this an error? Shouldn't it log one?

File: rotkehlchen/inquirer.py [hunk 15]
```diff
@@ -920,6 +927,68 @@ def find_curve_pool_price(
...
+        else:
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:969]:
```suggestion
        else:  # explain what this is in a short phrase
```

File: rotkehlchen/inquirer.py [hunk 16]
```diff
@@ -920,6 +927,68 @@ def find_curve_pool_price(
...
+            underlying_token = EvmToken(
+                evm_address_to_identifier(
+                    address=maybe_underlying_tokens[0].address,
+                    chain_id=token.chain_id,
+                    token_type=EvmTokenKind.ERC20,
+                ),
+            )
+
+        underlying_token_price = self.find_usd_price(underlying_token)
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:978]: This is only used in the else part ofthe try/except/else. Put it there and don't create a new variable for it.

Also ... can it raise anything? Is it handled?

File: rotkehlchen/types.py (hunk 3 above)
Comment (LefterisJP) [inline rotkehlchen/types.py:88]:
```suggestion
GEARBOX_PROTOCOL = 'gearbox'
```

**Review by LefterisJP — COMMENTED** [review]
Some comments.

Also sqldiff is just the one ABI you use right?

**Comment by moxw** [comment]
> Some comments.
> 
> Also sqldiff is just the one ABI you use right?

yes

File: rotkehlchen/inquirer.py (hunk 14 above)
Comment (moxw) [reply rotkehlchen/inquirer.py:968]: added

File: rotkehlchen/inquirer.py (hunk 16 above)
Comment (moxw) [reply rotkehlchen/inquirer.py:978]: From the docs, other usages of this function and looking at the code, it looks like the errors are handled in the function itself.  It instead `ZERO_PRICE` when it has exhausted all the options, so adding a error log if price == ZERO_PRICE.

**Review by LefterisJP — APPROVED** [review]
LGTM
